### PR TITLE
Feature/add cand active filter

### PR DESCRIPTION
**Pull request by qqss88**
## Summary (required)
'active_candidates' filter added to 'candidates/totals' api
- Resolves # [_Issue number_]
#3572 
_Include a summary of proposed changes._
'active_candidates' filter added to current api:

active_candidates=True(default):load active candidates only
active_candidates=False: load all candidates

files changed/added:
M       manage.py
M       webservices/args.py
M       webservices/common/models/candidates.py
M       webservices/resources/candidate_aggregates.py
M       webservices/rest.py
 A       webservices/resources/decorators.py
M       test_aggregates.py

## How to test the changes locally
* pull feature/add-cand-active-filter branch to your local environment
* Point SQLA_CONN to dev database
* run local server
* Open a browser window to http://localhost:5000 for testing:
load only active candidates
http://127.0.0.1:5000/v1/candidates/totals?api_key=NICAR16&active_candidates=True
or
http://127.0.0.1:5000/v1/candidates/totals?api_key=NICAR16
load all candidates
http://127.0.0.1:5000/v1/candidates/totals?api_key=NICAR16&active_candidates=False

test_aggregates.py updated to include auto testing for this.

## Impacted areas of the application
List general components of the application that this PR will affect:
https://www.fec.gov/data/raising-bythenumbers/
-  

## Related PRs
List related PRs against other branches:

branch | PR
------ | ------
fix/other_pr | [link]()
feature/other_pr | [link]()

### CODE REVIEW

File: webservices/args.py [hunk 1]
```diff
@@ -716,6 +716,7 @@ def make_seek_args(field=fields.Int, description=None):
     'max_debts_owed_by_committee': Currency(description='Maximum debt'),
     'federal_funds_flag': fields.Bool(description=docs.FEDERAL_FUNDS_FLAG),
     'has_raised_funds': fields.Bool(description=docs.HAS_RAISED_FUNDS),
+    'active_candidates': fields.Bool(missing=True, description=docs.ACTIVE_CANDIDATE),
```
Comment (lbeaufort) [inline webservices/args.py]: Defaulting to only show active candidates is a breaking change - we should keep the existing behavior the same and include all candidates (active and inactive) by default.

File: manage.py [hunk 2]
```diff
@@ -254,7 +254,7 @@ def cf_startup():
     """Migrate schemas on `cf push`."""
     check_config()
     if env.index == '0':
-        subprocess.Popen(['python', 'manage.py', 'refresh_materialized'])
+        subprocess.Popen(['python', '', 'refresh_materialized'])
```
Comment (lbeaufort) [inline manage.py]: I don't think you meant to delete this.

File: manage.py [hunk 3]
```diff
@@ -286,7 +286,7 @@ def load_efile_sheets():
 @manager.command
 def slack_message(message):
     """ Sends a message to the bots channel. you can add this command to ping you when a task is done, etc.
-    run ./manage.py slack_message 'The message you want to post'
+    run ./ slack_message 'The message you want to post'
```
Comment (lbeaufort) [inline manage.py]: I don't think you meant to delete this.

File: webservices/resources/candidate_aggregates.py [hunk 4]
```diff
@@ -165,11 +166,14 @@ def filter_range_fields(self, model):
         ('has_raised_funds', models.CandidateTotal.has_raised_funds),
         ('federal_funds_flag', models.CandidateTotal.federal_funds_flag),
         ('election_full', models.CandidateTotal.is_election),
+        # ('candidate_inactive', models.CandidateHistoryWithFuture.candidate_inactive),
```
Comment (lbeaufort) [inline webservices/resources/candidate_aggregates.py]: Why is this commented out? Should it be removed?

File: webservices/resources/candidate_aggregates.py [hunk 5]
```diff
@@ -165,11 +166,14 @@ def filter_range_fields(self, model):
...
     ]
 
+    #@print_query
     def build_query(self, **kwargs):
         history = models.CandidateHistoryWithFuture
         query = db.session.query(
+            #history
```
Comment (lbeaufort) [inline webservices/resources/candidate_aggregates.py]: Why is this commented out? Should it be removed?

File: webservices/resources/decorators.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+from functools import wraps
+from time import time
+import logging
+
+LOGGER = logging.getLogger(__name__)
+
+
+def timing(func):
```
Comment (lbeaufort) [inline webservices/resources/decorators.py]: Is this used?

File: webservices/resources/decorators.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        ts = time()
+        res = func(*args, **kwargs)
+        te = time()
+        LOGGER.info('STEP {} took {} seconds'.format(func.__name__, round(te - ts, 2)) + '\n')
+        return res
+    return wrapper
+
+# TODO: check this link to have an updated version of query print:
+# https://stackoverflow.com/questions/5631078/sqlalchemy-print-the-actual-query
+def print_query(func):
```
Comment (lbeaufort) [inline webservices/resources/decorators.py]: This @print_query function doesn't include the parameters, so I'm not sure it adds value - This feature needs more work and should probably go in another ticket for testing. To include parameters, you can do this, but it won't work for dates:
```
from sqlalchemy.dialects import postgresql

print(str(query.statement.compile(
    dialect=postgresql.dialect(),
    compile_kwargs={"literal_binds": True})))
```

**Comment by lbeaufort** [comment]
@qqss88 it looks like you have merge conflicts - can you rebase off develop?

File: webservices/resources/candidate_aggregates.py (hunk 5 above)
Comment (lbeaufort) [reply webservices/resources/candidate_aggregates.py]: Sorry if this is a comment - I thought it was commented out code. If it's a comment, I don't think it adds value here.

File: webservices/args.py (hunk 1 above)
Comment (qqss88) [reply webservices/args.py]: that is something we have discussed and confirmed with Paul - we'll keep the current change for now.. Thx

File: manage.py (hunk 2 above)
Comment (qqss88) [reply manage.py]: good catch - I think it is an accident. I rebased manage.py from develop branch. Thx.

File: manage.py (hunk 3 above)
Comment (qqss88) [reply manage.py]: same as above.

File: webservices/resources/candidate_aggregates.py (hunk 4 above)
Comment (qqss88) [reply webservices/resources/candidate_aggregates.py]: cleaned. Thx.

File: webservices/resources/candidate_aggregates.py (hunk 5 above)
Comment (qqss88) [reply webservices/resources/candidate_aggregates.py]: cleaned. thx.

File: webservices/resources/decorators.py (hunk 6 above)
Comment (qqss88) [reply webservices/resources/decorators.py]: I only used it for performance testing purpose. I'll remove it for now.

File: webservices/args.py (hunk 1 above)
Comment (lbeaufort) [reply webservices/args.py]: Thanks @qqss88. If this was a new feature, that might be fine, but the /candidates/totals/ endpoint feeds into the candidate datatables on the front end, and the endpoint has been around for a while. Changing expected behavior is a breaking change, so we shouldn't provide a default value and hide inactive candidates. I confirmed this with @PaulClark2.

**Comment by qqss88** [comment]
> It looks like the develop branch you merged into your branch may have been old? GitHub is showing 2,972 commits in this PR.

first time resolving conflicts using the web interface - not sure what is happened. I can rebase using the current version of dev branch if merging is an issue.

**Pull request closed without merging**